File: src/core/request.ts
import { invoke } from '@tauri-apps/api/core';

export interface RequestOptions {
    headers?: Record<string, string>;
    cookie?: string;
    proxyUrl?: string;
    fileName?: string;
}

export interface FetchWithHeadersResponse {
    text: string;
    headers?: Record<string, string>;
}

function normalizeProxyUrl(proxyUrl?: string): string | undefined {
    const value = proxyUrl?.trim();
    return value ? value : undefined;
}

function toErrorMessage(error: unknown): string {
    if (error instanceof Error) {
        return error.message;
    }
    return String(error);
}

export async function fetchText(url: string, options: RequestOptions = {}): Promise<string> {
    try {
        return await invoke<string>('fetch_data_command', {
            url,
            proxyUrl: normalizeProxyUrl(options.proxyUrl),
        });
    } catch (error) {
        throw new Error(`Request failed: ${toErrorMessage(error)}`);
    }
}

export async function postText(url: string, body: string, options: RequestOptions = {}): Promise<string> {
    try {
        return await invoke<string>('fetch_data_post', {
            url,
            body,
            proxyUrl: normalizeProxyUrl(options.proxyUrl),
        });
    } catch (error) {
        throw new Error(`Post request failed: ${toErrorMessage(error)}`);
    }
}

export async function fetchTextWithCookie(url: string, cookie: string, options: RequestOptions = {}): Promise<string> {
    try {
        return await invoke<string>('fetch_data_with_cookie', {
            url,
            cookie,
            proxyUrl: normalizeProxyUrl(options.proxyUrl),
        });
    } catch (error) {
        throw new Error(`Cookie request failed: ${toErrorMessage(error)}`);
    }
}

export async function fetchTextWithHeaders(
    url: string,
    headers: Record<string, string>,
    options: RequestOptions = {}
): Promise<FetchWithHeadersResponse> {
    try {
        return await invoke<FetchWithHeadersResponse>('fetch_data_with_headers_command', {
            url,
            headersJson: JSON.stringify(headers),
            proxyUrl: normalizeProxyUrl(options.proxyUrl),
        });
    } catch (error) {
        throw new Error(`Header request failed: ${toErrorMessage(error)}`);
    }
}

export async function postProtobuf(url: string, buffer: Uint8Array, options: RequestOptions = {}): Promise<string> {
    try {
        return await invoke<string>('fetch_data_buffer_base64', {
            url,
            buffer,
            proxyUrl: normalizeProxyUrl(options.proxyUrl),
            fileName: options.fileName ?? 'file',
        });
    } catch (error) {
        throw new Error(`Protobuf request failed: ${toErrorMessage(error)}`);
    }
}

export async function fetchData(url: string, options: RequestOptions = {}): Promise<string> {
    return fetchText(url, options);
}

export async function fetch_data_buffer(
    url: string,
    buffer: Uint8Array,
    options: RequestOptions = {}
): Promise<string> {
    return postProtobuf(url, buffer, options);
}

export async function fetchDataPost(url: string, body: string, options: RequestOptions = {}): Promise<string> {
    return postText(url, body, options);
}

export async function fetchData_with_cookie(
    url: string,
    cookie: string,
    options: RequestOptions = {}
): Promise<string> {
    return fetchTextWithCookie(url, cookie, options);
}

export async function fetch_data_with_headers_command(
    url: string,
    headers: Record<string, string>,
    options: RequestOptions = {}
): Promise<FetchWithHeadersResponse> {
    return fetchTextWithHeaders(url, headers, options);
}

File: src-tauri/src/request.rs
use reqwest::header::{HeaderMap, HeaderValue, CONTENT_TYPE, COOKIE};
use reqwest::{Client, Proxy};
use serde::Serialize;
use std::collections::HashMap;
use tauri::command;

#[derive(Serialize)]
pub struct ResponseData {
    pub text: String,
    pub headers: HashMap<String, String>,
}

fn build_client(proxy_url: Option<&str>, headers: Option<HeaderMap>) -> Result<Client, String> {
    let mut builder = Client::builder();

    if let Some(proxy_url) = proxy_url.filter(|value| !value.trim().is_empty()) {
        let proxy = Proxy::all(proxy_url).map_err(|error| format!("Invalid proxy: {}", error))?;
        builder = builder.proxy(proxy);
    }

    if let Some(headers) = headers {
        builder = builder.default_headers(headers);
    }

    builder
        .build()
        .map_err(|error| format!("Failed to build HTTP client: {}", error))
}

pub async fn fetch_data(url: &str, proxy_url: Option<&str>) -> Result<String, String> {
    let client = build_client(proxy_url, None)?;
    let response = client
        .get(url)
        .send()
        .await
        .map_err(|error| format!("Failed to send request: {}", error))?;

    response
        .text()
        .await
        .map_err(|error| format!("Failed to read response body: {}", error))
}

#[command]
pub async fn fetch_data_buffer(
    url: &str,
    buffer: Vec<u8>,
    file_name: &str,
    proxy_url: Option<String>,
) -> Result<Vec<u8>, String> {
    let client = build_client(proxy_url.as_deref(), None)?;
    let boundary = "-*_r1999";

    let mut body = Vec::new();
    body.extend_from_slice(format!("--{}\r\n", boundary).as_bytes());
    body.extend_from_slice(
        format!(
            "Content-Disposition: form-data; name=\"data\"; filename=\"{}\"\r\n\r\n",
            file_name
        )
        .as_bytes(),
    );
    body.extend_from_slice(&buffer);
    body.extend_from_slice(format!("\r\n--{}--\r\n", boundary).as_bytes());

    let mut headers = HeaderMap::new();
    headers.insert("Host", HeaderValue::from_static("tiebac.baidu.com"));
    headers.insert("User-Agent", HeaderValue::from_static("neotieba"));
    headers.insert("x_bd_data_type", HeaderValue::from_static("protobuf"));
    headers.insert("Connection", HeaderValue::from_static("keep-alive"));
    headers.insert("Accept", HeaderValue::from_static("*/*"));
    headers.insert(
        CONTENT_TYPE,
        HeaderValue::from_str(&format!("multipart/form-data; boundary={}", boundary))
            .map_err(|error| format!("Invalid content type: {}", error))?,
    );

    let response = client
        .post(url)
        .headers(headers)
        .body(body)
        .send()
        .await
        .map_err(|error| format!("Failed to send protobuf request: {}", error))?;

    if !response.status().is_success() {
        return Err(format!("Request failed with status: {}", response.status()));
    }

    response
        .bytes()
        .await
        .map(|data| data.to_vec())
        .map_err(|error| format!("Failed to read response body: {}", error))
}

#[command]
pub async fn fetch_data_post(
    url: &str,
    body: String,
    proxy_url: Option<String>,
) -> Result<String, String> {
    let client = build_client(proxy_url.as_deref(), None)?;
    let response = client
        .post(url)
        .body(body)
        .send()
        .await
        .map_err(|error| format!("Failed to send post request: {}", error))?;

    response
        .text()
        .await
        .map_err(|error| format!("Failed to read response body: {}", error))
}

pub async fn fetch_data_with_headers(
    url: &str,
    headers: HeaderMap,
    proxy_url: Option<&str>,
) -> Result<ResponseData, String> {
    let client = build_client(proxy_url, None)?;
    let response = client
        .get(url)
        .headers(headers)
        .send()
        .await
        .map_err(|error| format!("Failed to send header request: {}", error))?;

    let headers = response
        .headers()
        .iter()
        .map(|(key, value)| (key.to_string(), value.to_str().unwrap_or("").to_string()))
        .collect();

    let text = response
        .text()
        .await
        .map_err(|error| format!("Failed to read response body: {}", error))?;

    Ok(ResponseData { text, headers })
}

#[command]
pub async fn fetch_data_with_cookie(
    url: &str,
    cookie: &str,
    proxy_url: Option<String>,
) -> Result<String, String> {
    let mut headers = HeaderMap::new();

    if !cookie.trim().is_empty() {
        headers.insert(
            COOKIE,
            HeaderValue::from_str(cookie).map_err(|error| format!("Invalid cookie: {}", error))?,
        );
    }

    let client = build_client(proxy_url.as_deref(), Some(headers))?;
    let response = client
        .get(url)
        .send()
        .await
        .map_err(|error| format!("Failed to send cookie request: {}", error))?;

    if !response.status().is_success() {
        return Err(format!("Request failed with status: {}", response.status()));
    }

    response
        .text()
        .await
        .map_err(|error| format!("Failed to read response body: {}", error))
}
